<!DOCTYPE html>
<html>
	<head>
		<meta charset="utf-8" />
		<title>Jobim 2019</title>
		<style>
			body {
	font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
}
h1, h2, h3, h4 {
	font-weight: 400;
	margin-bottom: 0;
	margin-top: 0;
}
.remark-slide-content h1 {
	font-size: 3em;
	display: inline-block;
}

.remark-slide-content h2 {
	font-size: 2em;
	display: inline-block;
}

.remark-slide-content h3 {
	font-size: 1.5em;
	display: inline-block;
	margin-left: 30px
}

.remark-slide-content h4 {
	font-size: 1.2em;
	display: inline-block;
	margin-left: 30px
}

.footnote {
	position: absolute;
	bottom: 3em;
}
li p { line-height: 1.25em; }
.red { color: #fa0000; }
.large { font-size: 2em; }
a, a > code {
	color: rgb(249, 38, 114);
	text-decoration: none;
}
code {
	background: none repeat scroll 0 0 #F8F8FF;
  border: 1px solid #DEDEDE;
  border-radius: 3px 	;
  padding: 0 0.2em;
}
.remark-code, .remark-inline-code { font-family: "Bitstream Vera Sans Mono", "Courier", monospace; }
.remark-code-line-highlighted     { background-color: #373832; }
.pull-left {
	float: left;
	width: 47%;
}
.pull-right {
	float: right;
	width: 47%;
}
.pull-right ~ p {
	clear: both;
}
#slideshow .slide .content code {
	font-size: 0.8em;
}
#slideshow .slide .content pre code {
	font-size: 0.9em;
	padding: 15px;
}
.main-title, .title {
	background: #272822;
	color: #777872;
	text-shadow: 0 0 20px #333;
}
.title h1, .title h2, .main-title h1, .main-title h2 {
	color: #f3f3f3;
	line-height: 0.8em;
}
/* Custom */
.remark-code {
	display: block;
	padding: 0.5em;
}

.small-size {
    font-size: x-small;
}

 /* table 
table {
	border-collapse: collapse;
    width: 100%;
}

table td, table th {
    border: 1px solid #ddd;
    padding: 8px;
}

table tr:nth-child(even){background-color: #f2f2f2;}

table tr:hover {background-color: #ddd;}

table th {
    padding-top: 12px;
    padding-bottom: 12px;
    text-align: left;
    background-color: #4CAF50;
    color: white;
}
*/
/* line */

hr {
	color: #4CAF50;
}

/* images */

img{
	display: block;
    width: 100%;
	height: 100%;
}

/* iframe */

iframe {
    width: 100%;
	height: 550px;
}

.small-size {
    font-size: x-small;
}


.small-frame {
    height: 600px;
    width: 900px;

    transform: scale(0.8); 
    transform-origin: 0 0;
}

.grey {
    color: grey;
}
		</style>
	</head>
	<body>
		<textarea id="source">
# Jobim 2019
-------------------------

<br><br><br>
<center>
    <h3>
        AskOmics, a collaborative user-friendly interface for integrating datasets with references resources using semantic web technologies 
    </h3>    
</center>

<br><br><br>
<center>
<small>
    Xavier Garnier<sup>1</sup>, Anthony Bretaudeau<sup>1,2</sup>, Fabrice Legeai<sup>1,2</sup>, Anne Siegel<sup>1</sup> and Olivier Dameron<sup>1</sup>
</small>
</center>

<br>



<small>
1. Inria, Univ Rennes, CNRS, IRISA<br/ >
2. INRA, UMR IGEPP, BIPAA<br/ >

</small>

---

## Outline
---------

1. The nightmare of data: how to integrate project data with reference datasets

2. AskOmics as a solution

3. Demo

4. What's next?


---


## The nightmare of data
---------

--

Big Data: Datasets so **large** or **complex** that traditional data processing is inadequate.


- More than 1500 databases, lack of interoperability, multiple file format
- Need to integrate public data and project specific data
- Integration and query require technical skills and time

Need a user-friendly interface for data integration and querying

![data_deluge](images/data_deluge.png "data_deluge")

---

## The nightmare of data
### How to explore data
---------

![schema](images/schema.png "schema")

---


## The nightmare of data
### The semantic web
---------

- RDF (resource description framework) can well describe entities and relations

```
:gene rdf:type owl:Class .
:chromosome rdf:type owl:ObjectProperty .
:chromosome rdfs:domain :Gene .

:AT001 rdf:type :Gene ;
       rdfs:label "AT001" ;
       :chromosome :AT1 ;
       :end 40000 ;
       :organism :Arabidopsis_thaliana ;
       :start 100 .
```
--

- SPARQL (SPARQL protocol and query language) can be used to extract information

```
SELECT ?gene ?label
WHERE {
    ?gene rdf:type :Gene .
    ?gene rdfs:label ?label .
}

```

---


## The nightmare of data
### The semantic web
---------

![integration_query_schema](images/integration_query_schema.png "integration_query_schema")

---



## The nightmare of data
### The semantic web
---------

![lod_cloud](images/lod_cloud.png "lod_cloud")


---


## AskOmics as a solution
---------

--

AskOmics is a web software for data integration (references data and project specific data) and query using semantic web

- Convert multiple data format into rdf triples, and store them into a triplestore

- Query the rdf graph using a user-friendly interface

- Save, relaunch and share queries and results with other users (AskOmics is multi-users)

- AskOmics ecosystem: tools to generate AskOmics compliant files (AskoR, AuReMe), Interoperability with Galaxy


---

## Demo
### Input files
---------

--

- Genetic files
    * *Arabidopsis thaliana* (GFF, TAIR)
    * *Brassica napus* genes (GFF, BBIP platform)
    * Orthology relation (TSV, Chalhoub *et al*, 2014)

- Differential expression
    * Differential expression of *Brassica napus* genes between roots and leaves  (TSV, EdgeR + AskoR)

- Pathway
    * genes reaction and pathway data of *Brassica napus* (TSV, Metacyc + AuReMe)


![schema_small](images/schema_small.png "schema_small")

---


## Demo
---------


<br>
<center>
    <h3>
        C4EST PARTI POUR LA <a href="http://localhost:6543">D2MO</a> §§§
    </h3>    
</center>

![askomics_logo](images/askomics_logo.png "askomics_logo")


---


## Demo
### Biological questions
---------

- Which genes of *Brassica napus* are more strongly expressed in the roots than in the leaves?

--

- What are their orthologs in *Arabidospis thaliana*? Are they in a QTL?

--

- In which biological reactions are the genes obtained involved, and in which metabolic pathways are they involved?

---



## What's next?
### Ongoing work
---------

--
- Improve reproducibility and sharing functionalities

    - Save and share queries
    - Automated integration and query with the API


--

- Reach a larger user base
    - Offer a library of templates (TSV headers)
    - Provides query templates

--

- Extend query expressivity (and/or
)

--


- Support for multiple endpoints (FederatedQueryScaler, wimmics)

--


- Refactoring interface

    - API: Flask microframework (python)
    - Task queue: Celery (python)
    - Front: React library (JS)


---



## Collaboration
---------

Thanks to all AskOmics users:

- C Bettembourg – IGEPP aphids (D. Tagu)
- A. Evrard – IGEPP rapeseed (M. Jubault)
- M.Aite/C.Frioux/A.Siegel AUREME
- C.Bettembourg - Sanofi
- P. Leroux/F.Lecerf/S.Lagarrigue (Metachick, metabolome)
- IFB project (CIRAD/INRA) - Connecting AskOmics with ”SouthGreen” endpoint
- M. Rousseau/J.Lucas/J.Ferreira de Carvalho/S.Knosp - Rapeseed
- A. Sarniguet,J.Chappat - Rhysophere Colza
- M. Louarn Hematology
- M. Conan Seaweed
- L. Guillot-Cloarec Uniprot/NextProt
- S. Daval (INRA) - Meta-transcriptomics
- BIPAA (Arthropodes)/BBIP (Rapeseed) endpoint in production mode
- M. Wery these CIFRE - Sanofi
- M. Louarn these INSERM-INRIA
- J. Yon - Pegase, INRA
- M. Gonzalez - CGR, Universidad de Chile
- O. Chakoory/E. Forano - INRA
- G. Rabut - IGDR - protein-protein interactions
- P. Baudier/M. Aite - Drugs repositioning
- V. Henry - Neuromarkers


---

## Usefull links
---------

- Github repos
    - AskOmics: [askomics/askomics](https://github.com/askomics/askomics)
    - AskOmics3: [xgaia/flaskomics](https://github.com/xgaia/flaskomics)

- Docs:
    - AskOmics: [askomics.readthedocs.io](https://askomics.readthedocs.io)
    - AskOmics3: [flaskomics.readthedocs.io](https://flaskomics.readthedocs.io)




- Running instance: [askomics.genouest.org](https://askomics.genouest.org)

- This slides: [xgaia.github.io/askomics/jobim/2019](https://xgaia.github.io/askomics/jobim/2019)

- Contact me: [xavier.garnier@irisa.fr](mailto:xavier.garnier@irisa.fr)


		</textarea>
		<script src="../../../js/remark-latest.min.js"></script>
		<script src="issues.js"></script>

		<script>
			var slideshow = remark.create({
				highlightStyle: 'monokai',
			});
		</script>
		<script></script>
	</body>
</html>
