<!DOCTYPE html>
<html>
	<head>
		<meta charset="utf-8" />
		<title>DSparK</title>
		<style>
			body {
	font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
}
h1, h2, h3, h4 {
	font-weight: 400;
	margin-bottom: 0;
	margin-top: 0;
}
.remark-slide-content h1 {
	font-size: 3em;
	display: inline-block;
}

.remark-slide-content h2 {
	font-size: 2em;
	display: inline-block;
}

.remark-slide-content h3 {
	font-size: 1.5em;
	display: inline-block;
	margin-left: 30px
}

.remark-slide-content h4 {
	font-size: 1.2em;
	display: inline-block;
	margin-left: 30px
}

.footnote {
	position: absolute;
	bottom: 3em;
}
li p { line-height: 1.25em; }
.red { color: #fa0000; }
.large { font-size: 2em; }
a, a > code {
	color: rgb(249, 38, 114);
	text-decoration: none;
}
code {
	background: none repeat scroll 0 0 #F8F8FF;
  border: 1px solid #DEDEDE;
  border-radius: 3px 	;
  padding: 0 0.2em;
}
.remark-code, .remark-inline-code { font-family: "Bitstream Vera Sans Mono", "Courier", monospace; }
.remark-code-line-highlighted     { background-color: #373832; }
.pull-left {
	float: left;
	width: 47%;
}
.pull-right {
	float: right;
	width: 47%;
}
.pull-right ~ p {
	clear: both;
}
#slideshow .slide .content code {
	font-size: 0.8em;
}
#slideshow .slide .content pre code {
	font-size: 0.9em;
	padding: 15px;
}
.main-title, .title {
	background: #272822;
	color: #777872;
	text-shadow: 0 0 20px #333;
}
.title h1, .title h2, .main-title h1, .main-title h2 {
	color: #f3f3f3;
	line-height: 0.8em;
}
/* Custom */
.remark-code {
	display: block;
	padding: 0.5em;
}
 /* table */
table {
	border-collapse: collapse;
    width: 100%;
}

table td, table th {
    border: 1px solid #ddd;
    padding: 8px;
}

table tr:nth-child(even){background-color: #f2f2f2;}

table tr:hover {background-color: #ddd;}

table th {
    padding-top: 12px;
    padding-bottom: 12px;
    text-align: left;
    background-color: #4CAF50;
    color: white;
}

/* line */

hr {
	color: #4CAF50;
}

/* images */

img{
	display: block;
    width: 100%;
	height: 100%;
}

/* iframe */

iframe {
    width: 100%;
	height: 550px;
	zoom: 0.70;

}






		</style>
	</head>
	<body>
		<textarea id="source">
# DSparK
-------------------------

.footnote[Xavier Garnier, INRA]

---

## Table of contents
---------

1. dsk / GATB-core
2. DSparK
    - String kmers
    - Binary kmers
    - Dataframe

3. Benchmarks
    - Datasets

---


## dsk / GATB-core
------------------

DSK is a k-mer counter for reads or genomes. It use the GATB-core library (C++)

--
count: false

- input: set af sequences (fasta/q)
- output: set of kmers which occur more than a minimal amount of times, and number of times these kmers occur

output format is a binary file in h5 format. This output can converted to ascii with `dsk2ascii`

---
## dsk / GATB-core
------------------

A kmer and its reverse complement are considered to be the same kmer. DSK considers that A<C<T<G and returns the lexicographically smaller kmer using this alphabet order.


---
## DSparK
------------------


<b>D</b><b><i>S</i></b><i>par</i><b><i>K</i></b> is the spark implementation of dsk. It's available on the lipm GitLab.

- Code: https://lipm-gitlab.toulouse.inra.fr/spark-ics/DSparK
- Doc: https://lipm-gitlab.toulouse.inra.fr/spark-ics/DSparK/wikis/home


```bash
spark-submit --master yarn
             --deploy-mode cluster
             --num-executors 150
             --executor-memory 5G
             target/scala-2.11/DSparK-assembly-0.1.jar
                --input input_dir
                --kmer-size 31
                --abundance-max 2147483647
                --abundance-min 2
                --input-type fasta
                --output output_dir
```


* input: path to a fasta/q file, or a directory containing fasta/q files
* output: path to a directory that not exist

DSparK use [paramparser](https://lipm-gitlab.toulouse.inra.fr/llegrand/scala-utils)


---
## DSparK
### String kmers
------------------

First version of DSparK handle kmers in String Format


![dspark1](images/dspark1.png "DSparK 1")




---
## DSparK
### Binary kmers
-----------------

The second version handle kmers in binary format (Long)


![dspark2](images/dspark2.png "DSparK 2")


---
## DSparK
### Binary kmers
-----------------

DSparK encodes nucleotids the same way as GATB

```c
// GATB, C++
int nt2int(char nt) {
    return (nt >> 1 & 3);
}
```


```scala
// DSparK, Scala
def nuclToLong(nucl: Char): Long = {
    nucl >> 1 & 3
}
```


| Char | Int | Binary | nt2int (binary) | nt2int (int) |
|:---:|:---:|:---:|:---:|:---:|
| A | 65 | 1000<span style="color:red">00</span>1 | 00 | 0 |
| C | 67 | 1000<span style="color:red">01</span>1 | 01 | 1 |
| T | 84 | 1010<span style="color:red">10</span>0 | 10 | 2 |
| G | 71 | 1000<span style="color:red">11</span>1 | 11 | 3 |


---
## DSparK
### Binary kmers
-----------------

DSparK use a 64 bits `Long` to store the kmer. we can store a kmer with a maximum size of 32

Kmer and its revcomp are computed at the same time

The smallest `Long`  is the canonical one.




---
## DSparK
### Spark DataFrame
-----------------

A Dataframe is a distributed collection of data. It provides the benefits of RDDs (strong typing, ability to use powerful lambda functions) with the benefits of Spark SQLâ€™s optimized execution engine.

---
## DSparK
### Implementations
-----------------

4 implementation of DSparK:

- String Kmers witk RDD
- String kmers with DF
- Binary kmers with RDD
- Binary kmers with DF


---
## Benchmarks
### Datasets
-----------------
Input files: 4 fasta files:

|File|Size|Number of reads|
|:-----:|:-----:|:-----:|
|ERR194146_1.fasta|127 G|813 180 578|
|ERR194146_2.fasta|127 G|813 180 578|
|ERR194147_1.fasta|122 G|787 265 109|
|ERR194147_2.fasta|122 G|787 265 109|

3 datasets:

|Dataset|Files|Size|Number of reads|
|:-----:|:-----:|:-----:|:-----:|
|DS1|ERR194146_1.fasta|127 G|813 180 578|
|DS2|ERR194146_1.fasta + ERR194146_1.fasta|254 G|1 626 361 156|
|DS3|all|497 G|3 200 891 374|



---
## Benchmarks
### DSK
-----------------

<div id="benchmark_dsk_histogram"></div>



---
## Benchmarks
### DSparK
-----------------

Comparison of the 4 implementations of DSparK, on DS2, with 150 executors


<div id="benchmark_spark_histogram"></div>


---
## Benchmarks
### DSparK
------------

Execution time of DSparK on the 3 datasets, in function of the number of executors.

<div id="spark_benchmark"></div>

---
## Benchmarks
### Comparison
------------

Comparison between DSK and DSparK (on DS2):
- With the same ressources (40 cores/executors)
- The best execution time (40 cores for DSK, 150 executors for DSparK)

<div id="comparison_benchmark"></div>



---
## Conclusion
------------

- Execution time

With equivalente resources, it's difficult to compete a C++ program, but with Spark, we can easily access more ressources.

- Developpment time

Spark programs are easily to developp (once learning time is done)













		</textarea>
		<script src="../js/remark-latest.min.js"></script>
		<script src="../js/d3.v5.min.js"></script>
		<script type="text/javascript" src="../js/loader.js"></script>
		<script type="text/javascript" src="../js/fusioncharts.js"></script>
        <script type="text/javascript" src="graph.js"></script>
		<script>
			var slideshow = remark.create();
		</script>
		<script></script>
	</body>
</html>
